snuba/datasets/discover.py

```diff
@@ -0,0 +1,236 @@
+    """
+    Given a query, we attempt to guess whether it is better to fetch data from the
+    "events" or "transactions" dataset. This is going to be wrong in some cases.
+    """
+    # First check for a top level condition that matches either type = transaction
+    # type != transaction.
+    conditions = query.get_conditions()
+    if conditions:
+        for idx, condition in enumerate(conditions):
+            if is_condition(condition):
+                if tuple(condition) == ('type', '!=', 'transaction'):
+                    return EVENTS
+                elif tuple(condition) == ('type', '=', 'transaction'):
+                    return TRANSACTIONS
+
+    # If there is a condition that references a transactions only field, just switch
+    # to the transactions dataset
+    transaction_column_set = get_dataset('discover') \
+        .get_dataset_schemas() \
+        .get_read_schema() \
+        .get_transactions_only_columns()
+
+    if [col for col in query.get_columns_referenced_in_conditions() if transaction_column_set.get(col)]:
+        return TRANSACTIONS
+
+    # Use events by default
+    return EVENTS
+
+
+class DiscoverProcessor(QueryProcessor):
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        """
+        Switches the data source from the default (Events) to the transactions
+        table if a transaction specific column is detected.
+        Sets all the other columns to none
+
+        """
+        detected_dataset = detect_dataset(query)
+
+        source = get_dataset(detected_dataset) \
+            .get_dataset_schemas() \
+            .get_read_schema() \
+            .get_data_source()
+        query.set_data_source(source)
+
+
+class DiscoverSchema(Schema, ABC):
+    def get_data_source(self):
+        """
+        This is a placeholder, we switch out the data source in the processor
+        depending on the detected dataset
+        """
+        return None
+
+    def get_columns(self) -> ColumnSet:
+        common = ColumnSet([
+            ('event_id', FixedString(32)),
+            ('project_id', UInt(64)),
+            ('timestamp', DateTime()),
+            ('deleted', UInt(8)),
+            ('retention_days', UInt(16)),
+            ('platform', Nullable(String())),
+            ('sentry:release', Nullable(String())),
+            ('sentry:dist', Nullable(String())),
+            ('sentry:user', Nullable(String())),
+            # User
+            ('user_id', Nullable(String())),
+            ('username', Nullable(String())),
+            ('email', Nullable(String())),
+            ('ip_address', Nullable(String())),
+        ])
+
+        return common + self.get_events_only_columns() + self.get_transactions_only_columns()
+
+    def get_events_only_columns(self) -> ColumnSet:
+        return ColumnSet([
+            ('group_id', Nullable(UInt(64))),
+            ('primary_hash', Nullable(FixedString(32))),
+            ('type', Nullable(String())),
```

Comment:
Should we consider this common and auto-fill it with "transaction" for transactions ? Which also makes it non nullable.

Reply:
Having it synthesized for transactions would be great. Makes the result sets appear more consistent to the end user.

Reply:
Thanks made it not nullable. It's already being auto filled

Reply:
Actually, now I remember why it's nullable. It's because it's defined as nullable in Events. Changing it back.

Reply:
meh